sensor/common/sensor/central_communication_test.go

```diff
@@ -84,70 +64,78 @@ func Test_CentralCommunicationSuite(t *testing.T) {
-	componentMessages := make(chan *message.ExpiringMessage, 10)
-	comms := NewCentralCommunication(false, NewFakeSensorComponent(componentMessages))
-
-	reachable := concurrency.Flag{}
-	// This implicitly starts a goroutine
-	comms.Start(c.conn, &reachable, c.mockHandler, c.mockDetector)
+var _ central.SensorServiceClient = (*MockSensorServiceClient)(nil)
 
-	return componentMessages
+var centralSyncMessages = []*central.MsgToSensor{
+	debuggerMessage.SensorHello("00000000-0000-4000-A000-000000000000"),
+	debuggerMessage.ClusterConfig(),
+	debuggerMessage.PolicySync([]*storage.Policy{}),
+	debuggerMessage.BaselineSync([]*storage.ProcessBaseline{}),
 }
 
-func createConnectionAndStartServer(fakeCentral *centralDebug.FakeService) (*grpc.ClientConn, func()) {
-	buffer := 1024 * 1024
-	listener := bufconn.Listen(buffer)
-
-	fakeCentral.ServerPointer = grpc.NewServer()
-	central.RegisterSensorServiceServer(fakeCentral.ServerPointer, fakeCentral)
-
-	go func() {
-		utils.IgnoreError(func() error {
-			return fakeCentral.ServerPointer.Serve(listener)
-		})
-	}()
+func (c *centralCommunicationSuite) Test_StartCentralCommunication() {
+	// Create a fake SensorComponent
+	responsesC := make(chan *message.ExpiringMessage)
+	defer close(responsesC)
+	comm := NewCentralCommunication(false, NewFakeSensorComponent(responsesC))
 
-	conn, err := grpc.DialContext(context.Background(), "", grpc.WithContextDialer(func(ctx context.Context, s string) (net.Conn, error) {
-		return listener.Dial()
-	}), grpc.WithTransportCredentials(insecure.NewCredentials()))
+	reachable := concurrency.Flag{}
+	mockService := &MockSensorServiceClient{
+		connected: concurrency.NewSignal(),
+		client:    mocksClient.NewMockServiceCommunicateClient(c.controller),
+	}
 
-	if err != nil {
-		panic(err)
+	expectSyncMessages(centralSyncMessages, mockService)
+	ch := make(chan struct{})
+	mockService.client.EXPECT().Send(gomock.Any()).Times(1).DoAndReturn(func(msg *central.MsgFromSensor) error {
+		defer close(ch)
+		c.Assert().NotNil(msg.GetEvent().GetSynced())
+		return nil
+	})
+	// Start the go routine with the mocked client
+	go comm.(*centralCommunicationImpl).sendEvents(mockService, &reachable, c.mockHandler, c.mockDetector)
+	mockService.connected.Wait()
+	// Pretend that a component (listener) is sending the sync event
+	responsesC <- message.New(syncMessage())
+	select {
+	case <-ch:
+		break
+	case <-time.After(5 * time.Second):
+		c.Fail("timeout reached waiting for the sync event")
 	}
+}
 
-	closeF := func() {
-		utils.IgnoreError(listener.Close)
-		fakeCentral.ServerPointer.Stop()
+func expectSyncMessages(messages []*central.MsgToSensor, service *MockSensorServiceClient) {
+	key := strings.ToLower(centralsensor.SensorHelloMetadataKey)
+	md := metadata.MD{
+		key: []string{"true"},
+	}
+	service.client.EXPECT().Header().AnyTimes().Return(md, nil)
+	service.client.EXPECT().CloseSend().AnyTimes()
```

Comment:
Idea: we should add a test that tests connection teardown as well. Like:
```go
service.client.EXPECT().CloseSend().Times(1)
go comm.(*centralCommunicationImpl).sendEvents(mockService, &reachable, c.mockHandler, c.mockDetector)
mockService.connected.Wait()
comm.Stop(nil)
```

Reply:
I've tried to write a quick test like the one you suggested but I ran into data races with the waiting groups in the sender/receiver 😨  ([here](https://github.com/stackrox/stackrox/blob/c3da55cd1c9cd948ac637d809eb89b4ba7ef60be/sensor/common/sensor/central_receiver_impl.go#L38)). I'll investigate if the problem is in the tests or we have some nasty problem there.

Reply:
Race condition [fixed]([7287ccb](https://github.com/stackrox/stackrox/pull/8030/commits/7287ccb30580294aef7e73b880f6d59bf60841ad)). Now the test works without any problems.